### ISSUE TITLE
Entity::destroy() doesn't trigger ComponentRemovedEvent<>

### ISSUE THREAD

**Issue by c0ffeeartc** [alecthomas/entityx#107]
`Entity::destroy()` doesn't trigger `ComponentRemovedEvent<>`.
Is this by design?
If so, to replicate result - should I on `EntityDestroyedEvent` check `hasComponent<>()` for all possible components and remove them manually, or there is a better way?

**Comment by alecthomas** [comment]
This is not really deliberate, but is a result of how components are managed. The component pools would have to emit the `ComponentRemovedEvent<>` because that is the only location that knows what type is being removed.

The `Entity` and `ComponentHandle` classes would also have to be moved into a separate header so that `Pool.h` could include them.

**Comment by c0ffeeartc** [comment]
Will this change or should I use workarounds?

Here is a minimal test, btw:
```
TEST_CASE_METHOD(EntityManagerFixture, "TestComponentRemovedEventOnEntityDestroyed") {
  struct ComponentRemovedReceiver : public Receiver<ComponentRemovedReceiver> {
    void receive(const ComponentRemovedEvent<Direction> &event) {
      removed = true;
    }

    bool removed=false;
  };

  ComponentRemovedReceiver receiver;
  ev.subscribe<ComponentRemovedEvent<Direction>>(receiver);

  REQUIRE(!(receiver.removed));
  Entity e = em.create();
  e.assign<Direction>(1.0, 2.0);
  e.destroy();
  REQUIRE(receiver.removed);
}
```

**Comment by notlion** [comment]
This would be a really welcome addition. As it stands, I almost always need to subscribe to `EntityDestroyedEvent` when subscribed to a `ComponentRemovedEvent`. I'd love to avoid the extra `has_component` lookups on every Entity that is destroyed.

**Comment by c0ffeeartc** [comment]
I tried to make it properly but ended up using hacky solution.

Moved `Pool.h` classes to `Entity.h` above `class EntityManager`. 
Changed `Pool::destroy`:
```
   virtual void destroy(Entity entity, bool isCalledFromDestroyEntity, std::size_t n) override {
    assert(n < size_);
    if(isCalledFromDestroyEntity)
        entity.remove<T>();
    else
    {
        T *ptr = static_cast<T*>(get(n));
        ptr->~T();
    }
```

**Comment by zackthehuman** [comment]
I just submitted a PR that does basically what is described in this issue. It splits out `Entity` and `ComponentHandle` into their own headers, and then emits _removed_ events for each component belonging to an `Entity` before emitting the `EntityDestroyedEvent`.

**Comment by alecthomas** [comment]
Fixed in #129
